Added observers for ledger events

danishjafri88 opened the pull request:
#137  Added Observers

kylehickinson commented on BraveRewardsUI/Settings/Auto-Contribute/Details/AutoContributeDetailsViewController.swift:
```diff
@@ -6,7 +6,7 @@ import UIKit
 
 class AutoContributeDetailViewController: UIViewController {
-  
+  private var ledgerObs: LedgerObserver
```
Rather not short-form it :)

kylehickinson commented on BraveRewardsUI/Settings/Auto-Contribute/Details/AutoContributeDetailsViewController.swift:
```diff
@@ -31,9 +33,13 @@ class AutoContributeDetailViewController: UIViewController {
   }
   
+  deinit {
+    state.ledger.remove(ledgerObs)
```
Not necessary

kylehickinson commented on BraveRewardsUI/Settings/Tips/Details/TipsDetailViewController.swift:
```diff
@@ -6,7 +6,7 @@ import UIKit
 
 class TipsDetailViewController: UIViewController {
-  
+  private var ledgerObs: LedgerObserver
```
Feedback same as auto-contribute file

kylehickinson commented on BraveRewardsUI/Settings/Auto-Contribute/Details/AutoContributeDetailsViewController.swift:
```diff
@@ -382,3 +386,47 @@ extension AutoContributeDetailViewController {
+
+/// Ledger Observers
+extension AutoContributeDetailViewController {
+  func setObservers() {
```
Assuming this is only ever done once, lets call it `setupLedgerObservers`

Also lets have this setup in `init`, and add `guard viewIsLoaded else { return }` before any part of the methods require the view constructed.  In this case below its going to be at the top

kylehickinson commented on BraveRewardsUI/Settings/Auto-Contribute/Details/AutoContributeCell.swift:
```diff
@@ -48,6 +48,8 @@ class AutoContributeCell: UITableViewCell, TableViewReusable {
     siteStackView.addArrangedSubview(siteImageView)
     siteStackView.addArrangedSubview(siteNameLabel)
     
+    siteImageView.snp.removeConstraints()
```
What's this for?

danishjafri88 replied:
siteImageView's constructor takes a height which it uses to set a height constraint. But the imageview is placed inside a stackview and the stackview is inside a self sizing UITableviewcell. When it's drawn because of conflicting constraints, causes jerks. Removing this makes imageview draw relative to the stackview.

kylehickinson replied:
Ahhh okay cool didn't notice it was a custom image view

kylehickinson commented on BraveRewardsUI/Settings/Auto-Contribute/Details/AutoContributeDetailsViewController.swift:
```diff
@@ -382,3 +386,47 @@ extension AutoContributeDetailViewController {
+/// Ledger Observers
+extension AutoContributeDetailViewController {
+  func setObservers() {
+    ledgerObs.excludedSitesChanged = { key, exclude -> Void in
```
`[weak self]` since we are holding a ref to `ledgerObs` in self

kylehickinson commented on BraveRewardsUI/Tipping/TippingViewController.swift:
```diff
@@ -25,17 +25,18 @@ enum PublisherMediaType: String {
+    state.ledger.add(ledgerObserver)
     super.init(nibName: nil, bundle: nil)
-    
+    setupLedgerObservers()
```
We're going to handle showing confirmation based on callbacks added in #143. Unfortunately we won't want to wait for reconcile or observe recurring tips being added since we won't know when a reconcile will complete (could never happen) and 2. technically but likley never going to happen: other recurring tips could be added that are unrelated

Whether or not the tip goes out instantly, or its just added to pending contributions table, is dependent on a bunch of factors

So likely just going to take out all observer stuff in `TippingViewController`

danishjafri88 commented:
> Still crashing when removing multiple auto-contribute entries:
> 
> ```
> Invalid update: invalid number of rows in section 0.  The number of rows contained in an existing section after the update (5) must be equal to the number of rows contained in that section before the update (5), plus or minus the number of rows inserted or deleted from that section (1 inserted, 0 deleted) and plus or minus the number of rows moved into or out of that section (0 moved in, 0 moved out).
> ```
> 
> My steps that caused the crash:
> 
> 1. Restore All publishers (2), causing the restore all publishers row to go away
> 2. Scroll down (top section is out of view), exclude 1 publisher
> 3. Exclude a second publisher
> 
> Crash line: https://github.com/brave/brave-rewards-ios/pull/144/files#diff-33d19beafb04380e3ae8944a9e34df8eR420
> 
> **Note**: I rebased this branch so make sure you delete your local branch before pulling

Fixed and pushed.

danishjafri88 commented:
> Sorry its still weirdly glitchy when removing, lets just dump the animations and just `reloadData` for MVP. We can revisit later:
> 
> [rewards-ac-animation.mov.zip](https://github.com/brave/brave-rewards-ios/files/3576463/rewards-ac-animation.mov.zip)

Yeah, its because we are updating multiple things: 1 Delete, 1 insert (Different sections) and reload visible cells in section 1. Moving reloadData for now.

kylehickinson commented on BraveRewardsUI/Settings/Tips/Details/TipsDetailViewController.swift:
```diff
@@ -279,4 +283,26 @@ extension TipsDetailViewController {
+          !self.tipsList.contains { $0.rewardsCategory == .recurringTip } {
+          sections.insert(Section.summary.rawValue)
+        }
+        self.tipsView.tableView.reloadSections(sections, with: .automatic)
```
still animating this, I think the whole section here can be replaced right?

danishjafri88 replied:
This is a different controller, should I remove animation here too?

kylehickinson replied:
Yeah lets just do all of it

kylehickinson commented on BraveRewardsUI/Settings/Auto-Contribute/Details/AutoContributeDetailsViewController.swift:
```diff
@@ -382,3 +382,28 @@ extension AutoContributeDetailViewController {
+      case .all:
+        tableView.reloadData()
+      case .excluded:
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3, execute: {
```
Is this delay still needed without animations?

danishjafri88 replied:
Yes, the rewards db can take up some time to update the list.

kylehickinson commented on BraveRewardsUI/Settings/Wallet/WalletDetailsViewController.swift:
```diff
@@ -52,4 +54,30 @@ class WalletDetailsViewController: UIViewController, RewardsSummaryProtocol {
+        return
+      }
+      let rows = self.summaryRows.map({ row -> RowView in
+        row.isHidden = true
```
Why are you setting them all to hidden here?

danishjafri88 replied:
To animate an item in stack view we have to toggle isHidden

kylehickinson replied:
So you're hiding all the current summary rows… but then animating them hidden below to then animate them to visible after? While not accessing `summaryRows` again? I'm confused

kylehickinson replied:
Ah, I see, you're hiding the rows that are going to be set below. I see now. Is there a reason why you don't just do 

```
// Update rows but set them hidden to be animated
self.detailsView.activityView.rows = self.summaryRows.map { $0.isHidden = true; return $0 }
```

Would probably alleviate confusion

danishjafri88 replied:
Yeah will update :)

kylehickinson commented on BraveRewardsUI/Wallet/WalletViewController.swift:
```diff
@@ -523,4 +512,42 @@ extension WalletViewController {
+        return
+      }
+      let rows = self.summaryRows.map({ row -> RowView in
+        row.isHidden = true
```
Same question as above
